### PR TITLE
Make mamba use cache

### LINKED ISSUE #30849: Mamba: `use_cache` is not passed through in `prepare_inputs_for_generation`
Hi :)

I think that [use_cache](https://huggingface.co/docs/transformers/main/en/model_doc/mamba#transformers.MambaForCausalLM.forward.use_cache) is supposed to be passed through here as well:
https://github.com/huggingface/transformers/blob/9fd606dbdb392e76e66a4e4e852948ebf814dbb6/src/transformers/models/mamba/modeling_mamba.py#L634-L647

I noticed when I wanted to get the cache when using `model.generate`, but it was not there although I set `use_cache=True`.

Edit: Just saw that `GenerateDecoderOnlyOutput` would have to be adjusted as well. It would need to contain `cache_params` similarly to `past_key_values`. I don't know if it's okay for you to bloat that even more.
https://github.com/huggingface/transformers/blob/9fd606dbdb392e76e66a4e4e852948ebf814dbb6/src/transformers/generation/utils.py#L2244-L2251

Cheers and thanks for the great work!

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #30849. As it was reported in the issue, mamba was not using "use_cache" from generation kwargs. This PR adds it into kwargs and also returns the "cache_params", so that it can actually be used if `use_cache=True`

I tried to run Mamba test for generation locally, but it's failing for me even on `main` prob due to hardware differences (it has assertion on logits level)

### CODE REVIEW

File: src/transformers/models/mamba/modeling_mamba.py [hunk 1]
```diff
@@ -633,18 +633,30 @@ def _update_model_kwargs_for_generation(
...
+        input_ids,
+        inputs_embeds=None,
+        attention_mask=None,
+        use_cache=None,
+        past_key_values: Optional[MambaCache] = None,
+        **kwargs,
     ):
         # only last token for inputs_ids if the state is passed along.
-        if cache_params is not None:
+        if past_key_values is not None:
```
Comment (vasqu) [inline src/transformers/models/mamba/modeling_mamba.py:645]: I wouldn't rename the cache params to past kvs since mamba is not based on the attention mechanism making this a bit confusing. (The cache in this case saves past outputs of the ssm and conv1d layers.)

It's also a bit confusing in the generation utils but is kinda a necessity due to the code structure.

File: src/transformers/models/mamba/modeling_mamba.py [hunk 2]
```diff
@@ -633,18 +633,30 @@ def _update_model_kwargs_for_generation(
...
             model_inputs = {"inputs_embeds": inputs_embeds}
         else:
             model_inputs = {"input_ids": input_ids}
 
-        model_inputs["cache_params"] = cache_params
+        model_inputs.update(
+            {
+                "cache_params": past_key_values,
+                "use_cache": use_cache,
+                "attention_mask": attention_mask,
```
Comment (vasqu) [inline src/transformers/models/mamba/modeling_mamba.py:657]: Shouldn't be necessary since mamba ignores the attention mask.

**Comment by vasqu** [comment]
Hope you don't mind the nits :D Just a mamba enthusiast over here.

**Comment by vasqu** [comment]
Also tested the changes locally with commit a3cdff417bb3db94d5402dc009a07df97954a863 and it worked fine for `pytest tests/models/mamba -k "test_simple_generate" -s -vvvvv` (the logit levels test). Haven't tried the others.

(GPU: 3080ti-16GB, CPU: 12th Gen Intel(R) Core(TM) i7-12700H)

File: src/transformers/models/mamba/modeling_mamba.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/mamba/modeling_mamba.py:645]: Yes, I agree here but this is done for consistency, and following other models that use a different name for cache (like xlnet). 

I will leave this to @gante if we want to use the same naming for cache or not, as we're in the stage of refactoring generate :)

File: src/transformers/models/mamba/modeling_mamba.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/mamba/modeling_mamba.py:657]: Ah, didn't know that, then it's not needed

File: src/transformers/generation/utils.py [hunk 3]
```diff
@@ -632,6 +632,8 @@ def _extract_past_from_model_output(self, outputs: ModelOutput, standardize_cach
             past_key_values = outputs.mems
         elif "past_buckets_states" in outputs:
             past_key_values = outputs.past_buckets_states
+        elif "cache_params" in outputs:
+            past_key_values = outputs.cache_params
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:636]: IMO we should not need to update the generation utils, unless we find cache params to be general enough?

File: src/transformers/models/mamba/modeling_mamba.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/mamba/modeling_mamba.py:645]: ➕ on @vasqu , the older models like XLNET are not good precedents. 
Let's keep `cache_params`

File: src/transformers/models/mamba/modeling_mamba.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/mamba/modeling_mamba.py:657]: There is no attention in mamba indeed

File: tests/models/mamba/test_modeling_mamba.py [hunk 4]
```diff
@@ -447,10 +447,9 @@ def test_simple_generate(self, device):
 
         model = MambaForCausalLM.from_pretrained("state-spaces/mamba-130m-hf", torch_dtype=torch.float16)
         model.to(device)
-        model.config.use_cache = True
         input_ids = tokenizer("Hey how are you doing?", return_tensors="pt")["input_ids"].to(device)
 
-        out = model.generate(input_ids, do_sample=False, max_new_tokens=10)
+        out = model.generate(input_ids, do_sample=False, use_cache=True, max_new_tokens=10)
```
Comment (ArthurZucker) [inline tests/models/mamba/test_modeling_mamba.py:452]: it's true by default no?

**Review by ArthurZucker — COMMENTED** [review]
thanks

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:636]: We need to pass in updated "cache" into every forward call for the model to actually "use_cache". So we have to update `model_kwargs` in generation utils. CRIIW, as I am assuming from the code that Mamba caching works same as Transformer models

As you said below, if we don't want to follow XLNet style, I can use its own naming for cache. I think in that case we have to make changes in XLNet and other models with different cache names, to make generation utils consistent. Otherwise we'll be using forcefully `past_key_values` for several models except Mamba, thus adding unnecessary conditional checks

File: tests/models/mamba/test_modeling_mamba.py (hunk 4 above)
Comment (zucchini-nlp) [reply tests/models/mamba/test_modeling_mamba.py:452]: I removed `model.config.use_cache = True` just above this and moved it into the `generate`, to test that current changes are working and generate accepts "use_cache" in kwargs for Mamba. 

The real changes are the ones in `prepare_inputs_for_generation` which didn't pass `use_cache` in kwargs earlier

File: src/transformers/generation/utils.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/generation/utils.py:636]: No, I think we ought to just leave previous model as is (they are using past_key_values) and just support arbitrary kwargs being returned from the model).

**Review by ArthurZucker — APPROVED** [review]
🔥 Thanks for updating

**Comment by ArthurZucker** [comment]
Forgot to ask about benchmarking the generation speed btw, does it improve it (because now it's actually usingthe cache) ?

**Comment by zucchini-nlp** [comment]
@ArthurZucker also forgot, will do a small evaluation to verify tomorrow :)

**Comment by ylacombe** [comment]
Out of curiosity, don't we have the same issue with Jamba?

**Comment by zucchini-nlp** [comment]
@ylacombe hey! I just checked jamba and it seems to be passing the updated cache every step, so there should be no such problem. But let me know if you noticed something else :)

**Comment by zucchini-nlp** [comment]
@ArthurZucker done! The evaluation shows that caching is faster and more memory efficient than without cache. Below are the results:

Peak at: 27 MiB for cached generation
Time taken: 28.82830 s for cached generation

Peak at: 532 MiB for generation wo cache
Time taken: 829.49494 s for cached generation

```python
from transformers import AutoTokenizer, MambaForCausalLM
import torch
from time import perf_counter

model_id = "state-spaces/mamba-2.8b-hf"
tokenizer = AutoTokenizer.from_pretrained(model_id)
tokenizer.pad_token = tokenizer.eos_token
model = MambaForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16, device_map="auto")
input_ids = tokenizer("Hey how are you doing?", return_tensors="pt")["input_ids"].to(model.device)

# Get init mem used up by model params
init_mem = torch.cuda.memory_allocated()
torch.cuda.reset_max_memory_allocated()

# Eval for cached generation
start = perf_counter()
out = model.generate(input_ids, do_sample=False, use_cache=True, max_new_tokens=500)
torch.cuda.synchronize()

total_time = perf_counter() - start
peak = (torch.cuda.max_memory_allocated() - init_mem) // 1024 ** 2
print(f"Peak at: {peak} MiB for cached generation")
print(f"Time taken: {total_time:.05f} s for cached generation")

# Reset and start eval on use_cache=False
torch.cuda.reset_max_memory_allocated()

start = perf_counter()
out = model.generate(input_ids, do_sample=False, use_cache=False, max_new_tokens=500)
torch.cuda.synchronize()

total_time = perf_counter() - start
peak = (torch.cuda.max_memory_allocated() - init_mem) // 1024 ** 2
print(f"Peak at: {peak} MiB for generation wo cache")
print(f"Time taken: {total_time:.05f} s for cached generation")
```

**Comment by ArthurZucker** [comment]
💘 great work
